backend/app/websocket.py

```diff
@@ -101,92 +108,157 @@ def process_chat_input(
-                    ConnectionId=connection_id, Data=data_to_send
+            elif isinstance(event, MessageStopEvent):
+                # Persist conversation before finish streaming so that front-end can avoid 404 issue
+                concatenated = "".join(completions)
+                # Append entire completion as the last message
+                assistant_msg_id = str(ULID())
+                message = MessageModel(
+                    role="assistant",
+                    content=[
+                        ContentModel(
+                            content_type="text", body=concatenated, media_type=None
+                        )
+                    ],
+                    model=chat_input.message.model,
+                    children=[],
+                    parent=user_msg_id,
+                    create_time=get_current_time(),
                 )
-            except Exception as e:
-                logger.error(f"Failed to post message: {str(e)}")
-                return {
-                    "statusCode": 500,
-                    "body": "Failed to send message to connection.",
-                }
-        elif isinstance(event, MessageDeltaEvent):
-            logger.debug(f"Received message delta event: {event.delta}")
-            last_data_to_send = json.dumps(
-                dict(
-                    completion="",
-                    stop_reason=event.delta.stop_reason,
+                conversation.message_map[assistant_msg_id] = message
+                # Append children to parent
+                conversation.message_map[user_msg_id].children.append(assistant_msg_id)
+                conversation.last_message_id = assistant_msg_id
+
+                # Update total pricing
+                metrics = event.model_dump()["amazon-bedrock-invocationMetrics"]
+                input_token_count = metrics.get("inputTokenCount")
+                output_token_count = metrics.get("outputTokenCount")
+
+                logger.debug(
+                    f"Input token count: {input_token_count}, output token count: {output_token_count}"
                 )
-            ).encode("utf-8")
-        elif isinstance(event, MessageStopEvent):
-            # Persist conversation before finish streaming so that front-end can avoid 404 issue
-            concatenated = "".join(completions)
-            # Append entire completion as the last message
-            assistant_msg_id = str(ULID())
-            message = MessageModel(
-                role="assistant",
-                content=[
-                    ContentModel(
-                        content_type="text", body=concatenated, media_type=None
+
+                price = calculate_price(
+                    chat_input.message.model, input_token_count, output_token_count
+                )
+                conversation.total_price += price
+
+                store_conversation(user_id, conversation)
+            else:
+                continue
+    else:
```

Comment:
It's a little bit long to read. Could you refactor this part? (the part of `for event in response ....`)

Reply:
similar above, right now keep both an anthropic models and the bedrock client streaming response parsing. Still think it's better to have another PR later to focus on refactoring.

Reply:
To keep readability, stream response handler should be declared such as following:

```py
def handle_stream_response(response):
   if is_anthropic:
      _handle_stream_anthropic_response(response)
    else:
      # MIstral
      _handle_stream_bedrock_response(response)
```

Reply:
@statefb ideally should like that. but currently the streaming response handling need a lot of extra context, conversation, connection and gatewayapi etc. 
In fact the `handle_stream_response` interface would look something like
```
def handle_stream_response(args, conversation, gatewayapi, connection_id, response):
....
```
in this case, it has not much difference with current `process_chat_input` function. I would suggest only if you have a way to optimize this flow, then the rectoring on this could work.

Reply:
@hustshawn I understood, makes sense.